metadata-ingestion/src/datahub/ingestion/source/s3/source.py

```diff
@@ -425,6 +426,13 @@ def create_container_hierarchy(
             if base_full_path.rfind("/") != -1
             else ""
         )
+
+        # Dataset is in the root folder
+        if not parent_folder_path:
+            if parent_key:
+                yield from add_dataset_to_container(parent_key, dataset_urn)
```

Comment:
Extract these lines out & refactor the following logic from the bottom of this function up here.
```
if parent_key is None:
            logger.warning(
                f"Failed to associate Dataset ({dataset_urn}) with container"
            )
            return
        yield from add_dataset_to_container(parent_key, dataset_urn)
```